File: 0907-koko-eating-bananas/solution.cs
public class Solution {
    public int MinEatingSpeed(int[] piles, int h) {
        int l = 1, r = piles.Max();
        int minK = l;
        while(l <= r){
            int m = l + (r - l) / 2;
            long currTime = 0;
            foreach(int p in piles){
                currTime += (int)Math.Ceiling((double)p / m);
            }
            if(currTime <= h){
                minK = m;
                r = --m;
            }
            else{
                l = ++m; 
            }
        }
        
        return minK;
    }
}

File: 0907-koko-eating-bananas/solution.py
from typing import List
import math

class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        low, high = 1, max(piles)
        ans = high
        while low <= high:
            mid = (low + high) // 2
            # Calculate hours needed at speed = mid
            curr = sum(math.ceil(pile / mid) for pile in piles)
            
            if curr <= h:
                ans = mid
                high = mid - 1  # try slower speed
            else:
                low = mid + 1   # need faster speed
        return ans

